File: 백준/Gold/16565. N포커/N포커.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        int MOD = 10007;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] dp = new int[53][53];
        int result = 0;
        
        dp[0][0] = 1;
        for (int i = 1; i < 53; i++) {
            dp[i][0] = dp[i][i] = 1;
            dp[i][1] = dp[i][i - 1] = i;
            for (int j =2; j <= i/2; j++) {
                dp[i][j] = (dp[i-1][j-1] + dp[i-1][j]) % MOD;
                dp[i][i-j] = dp[i][j];
            }
        }

        for (int i = 1; i <= n/4; i++) {
            if (i % 2 == 1) {
                result += dp[13][i] * dp[52-i*4][n-i*4] % MOD;
                result %= MOD;
            } else {
                result = (result - dp[13][i] * dp[52-i*4][n-i*4] % MOD + MOD) % MOD;
            }
        }
        System.out.println(result % MOD);
    }
}

File: 백준/Gold/16565. N포커/N포커.py
# N포커
import sys
input = sys.stdin.readline

MOD = 10007
n = int(input())

dp = [[0] * 53 for _ in range(53)]
dp[0][0] = 1

for i in range(1, 53):
    dp[i][0] = 1
    dp[i][i] = 1
    dp[i][1] = i
    for j in range(1, i//2+1):
        dp[i][j] = (dp[i-1][j] + dp[i-1][j-1]) % MOD
        dp[i][i-j] = dp[i][j]

result = 0
for i in range(1, i//4+1):
    if i % 2 == 1:
        result += dp[13][i] * dp[52-i*4][n-i*4] % MOD
    else:
        result -= dp[13][i] * dp[52-i*4][n-i*4] % MOD

print(result % MOD)
